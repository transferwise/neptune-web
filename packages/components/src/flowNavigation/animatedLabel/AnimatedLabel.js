import React from 'react';
import Types from 'prop-types';
import classNames from 'classnames';

import './AnimatedLabel.css';

const AnimatedLabel = ({ activeLabel, className, labels }) => {
  const numLabels = labels.length - 1;

  return (
    <div className={classNames('np-animated-label', className)}>
      {labels.map((label, index) => {
        const nextLabel = index - 1;
        return (
          <div
            // eslint-disable-next-line
            key={index}
            className={classNames('text-xs-left', {
              'np-animated-label--in': index === activeLabel,
              'np-animated-label--out': nextLabel === activeLabel && nextLabel !== numLabels,
            })}
          >
            {label}
          </div>
        );
      })}
    </div>
  );
};

AnimatedLabel.propTypes = {
  activeLabel: Types.number.isRequired,
  className: Types.string,
  labels: Types.arrayOf(Types.node).isRequired,
};
AnimatedLabel.defaultProps = {
  className: undefined,
};

export default AnimatedLabel;
