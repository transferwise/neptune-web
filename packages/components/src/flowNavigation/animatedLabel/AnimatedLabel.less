@import (reference) '@transferwise/neptune-css/src/less/addons/_spacing-utilities.less';

@transition-timing-function: ease-in-out;
@transition-duration: 0.3s;
@short-transition-duration: 0.15s;
@label-height: 24px;
@slide-length: 8px;

.np-animated-label {
  height: @label-height;
  overflow: hidden;
  position: relative;
  padding-top: 1px;
  font-weight: 600;

  > * {
    height: 0;
    opacity: 0;
    position: absolute;
    transform: translateX(-@slide-length);
    transition: all @transition-duration ease-in;
  }

  &--in {
    height: auto;
    opacity: 1;
    position: relative;
    top: auto;
    transform: translateX(0);
    transition: all @transition-duration ease-in @transition-duration;
  }

  &--out {
    position: relative;
    top: -@label-height;
    transform: translateX(@slide-length);
  }
}
