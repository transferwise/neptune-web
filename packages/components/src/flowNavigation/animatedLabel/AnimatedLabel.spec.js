import React from 'react';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

import AnimatedLabel from '.';

const props = {
  activeLabel: 0,
  labels: ['label1', 'label2', 'label3'],
};
jest.useFakeTimers();
describe('AnimatedLabel', () => {
  it(`renders all labels`, () => {
    const { container } = render(<AnimatedLabel {...props} />);
    expect(container).toMatchSnapshot();
  });

  it(`renders only one label with class in`, () => {
    const { container, getByText } = render(<AnimatedLabel {...props} />);
    expect(getByText(props.labels[0])).toHaveClass('np-animated-label--in');
    expect(container.querySelectorAll('.np-animated-label--in')).toHaveLength(1);
  });

  it(`renders only one label with class out`, () => {
    const { container, getByText } = render(<AnimatedLabel {...props} />);
    expect(getByText(props.labels[1])).toHaveClass('np-animated-label--out');
    expect(container.querySelectorAll('.np-animated-label--out')).toHaveLength(1);
  });

  it(`when activeLabel increase it switches class accordingly`, () => {
    const { getByText, rerender } = render(<AnimatedLabel {...props} />);
    expect(getByText(props.labels[0])).toHaveClass('np-animated-label--in');
    expect(getByText(props.labels[1])).toHaveClass('np-animated-label--out');

    expect(getByText(props.labels[2])).not.toHaveClass('np-animated-label--out');
    expect(getByText(props.labels[2])).not.toHaveClass('np-animated-label--in');

    rerender(<AnimatedLabel {...props} activeLabel={1} />);

    expect(getByText(props.labels[0])).not.toHaveClass('np-animated-label--out');
    expect(getByText(props.labels[0])).not.toHaveClass('np-animated-label--in');

    expect(getByText(props.labels[1])).toHaveClass('np-animated-label--in');
    expect(getByText(props.labels[2])).toHaveClass('np-animated-label--out');
  });

  it(`when activeLabel decrease it switches class accordingly`, () => {
    const { getByText, rerender } = render(<AnimatedLabel {...props} activeLabel={1} />);
    expect(getByText(props.labels[1])).toHaveClass('np-animated-label--in');
    expect(getByText(props.labels[2])).toHaveClass('np-animated-label--out');

    expect(getByText(props.labels[0])).not.toHaveClass('np-animated-label--out');
    expect(getByText(props.labels[0])).not.toHaveClass('np-animated-label--in');

    rerender(<AnimatedLabel {...props} activeLabel={0} />);

    expect(getByText(props.labels[0])).toHaveClass('np-animated-label--in');
    expect(getByText(props.labels[1])).toHaveClass('np-animated-label--out');

    expect(getByText(props.labels[2])).not.toHaveClass('np-animated-label--out');
    expect(getByText(props.labels[2])).not.toHaveClass('np-animated-label--in');
  });
});
