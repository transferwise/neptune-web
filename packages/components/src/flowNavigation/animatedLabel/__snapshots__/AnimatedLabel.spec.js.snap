// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedLabel renders all labels 1`] = `
<div>
  <div
    class="np-animated-label"
  >
    <div
      class="text-xs-left np-animated-label--in"
    >
      label1
    </div>
    <div
      class="text-xs-left np-animated-label--out"
    >
      label2
    </div>
    <div
      class="text-xs-left"
    >
      label3
    </div>
  </div>
</div>
`;
